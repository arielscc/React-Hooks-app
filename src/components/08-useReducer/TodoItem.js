import React from 'react';
import PropTypes from 'prop-types';

const TodoItem = React.memo(({ todo, index, handleToggle, handleDelete }) => {
  return (
    <li
      key={todo.id}
      onClick={() => handleToggle(todo.id)}
      className={`flex justify-between items-center mt-2 w-1/2 p-3 rounded-md cursor-pointer box-content  ${
        index % 2 === 0 && 'bg-gray-200'
      } ${
        todo.done &&
        'line-through bg-gradient-to-r from-indigo-500 to-indigo-300'
      }`}
    >
      <p>{todo.todo}</p>
      <button
        onClick={() => handleDelete(todo.id)}
        className="py-1 px-2 rounded-md bg-red-600 font-bold text-white focus:outline-none"
      >
        <svg
          className="w-6 h-6"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
            d="M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16"
          />
        </svg>
      </button>
    </li>
  );
});

TodoItem.propTypes = {
  todo: PropTypes.object.isRequired,
  index: PropTypes.number.isRequired,
  handleToggle: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
};

export default TodoItem;
