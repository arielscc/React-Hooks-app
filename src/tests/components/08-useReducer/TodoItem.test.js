import React from 'react';
import { shallow } from 'enzyme';
import TodoItem from '../../../components/08-useReducer/TodoItem';
import { demoTodos } from '../../fixtures/demoTodos';
import '@testing-library/jest-dom';

const handleDelete = jest.fn();
const handleToggle = jest.fn();

describe('Pruebas del componente TodoItem', () => {
  const wrapper = shallow(
    <TodoItem
      todo={demoTodos[1]}
      index={1}
      handleDelete={handleDelete}
      handleToggle={handleToggle}
    />
  );

  test('Debe hacer match con el snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('Debe de llamar la funcion handleDelete', () => {
    wrapper.find('button').simulate('click');
    expect(handleDelete).toHaveBeenCalledTimes(1);
    expect(handleDelete).toHaveBeenCalledWith(demoTodos[1].id);
  });

  test('Debe de llamar la funcion handleToggle', () => {
    wrapper.find('li').simulate('click');
    expect(handleToggle).toHaveBeenCalledTimes(1);
    expect(handleToggle).toHaveBeenCalledWith(demoTodos[1].id);
  });

  test('Debe de Mostrar el texto del parrafo correctamente', () => {
    expect(wrapper.find('p').text().trim()).toBe(demoTodos[1].todo);
  });

  test('Debe de cambiar el estilo del todo cuando se le hace click', () => {
    const todo = demoTodos[0];
    todo.done = true;

    const wrapper = shallow(
      <TodoItem
        todo={todo}
        index={1}
        handleDelete={handleDelete}
        handleToggle={handleToggle}
      />
    );
    wrapper.find('li').simulate('click');

    const todoLi = wrapper.hasClass('line-through');
    expect(todoLi).toBe(true);
  });
});
