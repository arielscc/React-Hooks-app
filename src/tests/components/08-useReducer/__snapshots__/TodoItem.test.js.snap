// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del componente TodoItem Debe hacer match con el snapshot 1`] = `
<li
  className="flex justify-between items-center mt-2 w-1/2 p-3 rounded-md cursor-pointer box-content  false false"
  key="2"
  onClick={[Function]}
>
  <p>
    Aprender Mongo
  </p>
  <button
    className="py-1 px-2 rounded-md bg-red-600 font-bold text-white focus:outline-none"
    onClick={[Function]}
  >
    <svg
      className="w-6 h-6"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
      />
    </svg>
  </button>
</li>
`;
